<a href="https://colab.research.google.com/github/hyerin0113/tobigs14_assign/blob/master/Week3_Naive_Bayes_Classifier/week3(NaiveBayes)_assignment_hyerinlee.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Assignment_투빅스 14기 이혜린

## 우리의 목표 : 나이브베이즈 문제 해결하기
1.1) 입력문서가 {fast, furious, fun} 만을 주요 단어로 가질때, 이 문서는 얼마의 확률로 어떤 문서로 분류되는가?

##### 노트북 파일을 따라오면서, 빈칸과 질문에 대한 물음을 모두 채우시면 과제 완료 입니다 : )

### 1. 주어진 dataset 을 확인해 보기위해 엑셀 파일을 읽습니다.

In [2]:
import numpy as np
import pandas as pd

In [4]:
documents = pd.read_excel('/content/drive/My Drive/Tobigs14/Week3_Naive_Bayes/test_file.xlsx') 

In [5]:
documents 

label                        message
0  comedy          Fun couple, love love
1  action          fast Furious, shoot!!
2  comedy  couple^^, fly, fast, fun, fun
3  action       furious, shoot shoot fun
4  action         fly~~ fast shoot, Love

### 2. 파일에 어떤 문제점이 보이시나요? (어떤 부분을 전처리 해주어야 할까요?)

####  Q1. 발견한 문제점(전처리가 필요한 부분)을 2가지 이상 적어주세요.
- 1. 대문자 혹은 소문자로 문자를 통일시켜야 한다.
- 2. 특수문자를 제거해야 한다.

### step 1. 단어를 모두 소문자로 바꾸어 줍니다. 
#### Q2.왜 이 과정이 필요할까요? 
- A. 대소문자의 차이로 인해 같은 단어가 다른 의미의 단어로 분류될수도 있기 때문이다.

#### Q3. 코드속 빈칸을 채워주세요

In [6]:
lower_documents = []
lower_documents = [d.lower() for d in documents['message']]
print(lower_documents)

['fun couple, love love', 'fast furious, shoot!!', 'couple^^, fly, fast, fun, fun', 'furious, shoot shoot fun', 'fly~~ fast shoot, love']


### step2. 특수문자를 없애줍니다.
#### string.punctuation 에는 특수문자들이 저장되어있는것을 확인할수 있습니다.

In [7]:
import string
string.punctuation

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

#### Q4. string.punctuation 을 이용해서 특수문자를 제거해주세요.

In [8]:
import re

del_punctuation_documents = []

for i in lower_documents:
    del_punctuation_documents.append(re.sub('['+string.punctuation+']', '', i))
    
del_punctuation_documents

['fun couple love love',
 'fast furious shoot',
 'couple fly fast fun fun',
 'furious shoot shoot fun',
 'fly fast shoot love']

### step 3. 단어를 하나씩 띄어쓰기 단위로 쪼개어 줍니다.
#### Q5. 빈칸을 채워주세요.

In [9]:
preprocessed_documents = [[w for w in d.split()] for d in del_punctuation_documents]
preprocessed_documents

[['fun', 'couple', 'love', 'love'],
 ['fast', 'furious', 'shoot'],
 ['couple', 'fly', 'fast', 'fun', 'fun'],
 ['furious', 'shoot', 'shoot', 'fun'],
 ['fly', 'fast', 'shoot', 'love']]

### step 4. 각각의 단어가 문서에서 몇번 나왔는지 세봅시다.

In [10]:
frequency_list = []
from collections import Counter

frequency_list = [Counter(d) for d in preprocessed_documents]
frequency_list

[Counter({'couple': 1, 'fun': 1, 'love': 2}),
 Counter({'fast': 1, 'furious': 1, 'shoot': 1}),
 Counter({'couple': 1, 'fast': 1, 'fly': 1, 'fun': 2}),
 Counter({'fun': 1, 'furious': 1, 'shoot': 2}),
 Counter({'fast': 1, 'fly': 1, 'love': 1, 'shoot': 1})]

### step 5. 이제 우리가 하고싶은것은, 문자를 숫자로 변환하는 작업입니다!(그림 참고)
- 이를 위해 CountVectorizer를 사용합니다.
- CountVectorizer는 문서 집합에서 단어 토큰을 생성하고, 각 단어 수를 세어 BOW (Bag of Words) 인코딩한 벡터를 만들어줍니다.

- 문서에 해당단어가 몇번 포함되었는지를 나타낼 때 사용하는 방법입니다.

- 참고 : https://datascienceschool.net/view-notebook/3e7aadbf88ed4f0d87a76f9ddc925d69/

![image.png](attachment:image.png)

In [11]:
from sklearn.feature_extraction.text import CountVectorizer
count_vector = CountVectorizer()
count_vector.fit(documents['message'])

CountVectorizer(analyzer='word', binary=False, decode_error='strict',
                dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
                lowercase=True, max_df=1.0, max_features=None, min_df=1,
                ngram_range=(1, 1), preprocessor=None, stop_words=None,
                strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
                tokenizer=None, vocabulary=None)

In [12]:
doc_array = count_vector.transform(documents['message']).toarray()
doc_array

array([[1, 0, 0, 1, 0, 2, 0],
       [0, 1, 0, 0, 1, 0, 1],
       [1, 1, 1, 2, 0, 0, 0],
       [0, 0, 0, 1, 1, 0, 2],
       [0, 1, 1, 0, 0, 1, 1]])

In [13]:
count_vector.get_feature_names()

['couple', 'fast', 'fly', 'fun', 'furious', 'love', 'shoot']

In [14]:
count_vector.vocabulary_ # index 설명

{'couple': 0,
 'fast': 1,
 'fly': 2,
 'fun': 3,
 'furious': 4,
 'love': 5,
 'shoot': 6}

In [15]:
frequency_matrix = pd.DataFrame(doc_array, columns = count_vector.get_feature_names())
frequency_matrix

couple  fast  fly  fun  furious  love  shoot
0       1     0    0    1        0     2      0
1       0     1    0    0        1     0      1
2       1     1    1    2        0     0      0
3       0     0    0    1        1     0      2
4       0     1    1    0        0     1      1

In [16]:
frequency_matrix['count'] = frequency_matrix.sum(axis=1)

In [17]:
frequency_matrix

couple  fast  fly  fun  furious  love  shoot  count
0       1     0    0    1        0     2      0      4
1       0     1    0    0        1     0      1      3
2       1     1    1    2        0     0      0      5
3       0     0    0    1        1     0      2      4
4       0     1    1    0        0     1      1      4

### step 6. 범주형 변수를 dummy변수로 변환해주는 작업(One-Hot Encoding!)을 해주어야합니다.
#### Q6. label 을 comedy =1, action =0 으로 변환해주세요

In [18]:
documents["label"]=pd.get_dummies(documents['label'], drop_first=True)

In [19]:
doc = pd.concat([documents['label'],frequency_matrix],axis=1)

In [20]:
doc

label  couple  fast  fly  fun  furious  love  shoot  count
0      1       1     0    0    1        0     2      0      4
1      0       0     1    0    0        1     0      1      3
2      1       1     1    1    2        0     0      0      5
3      0       0     0    0    1        1     0      2      4
4      0       0     1    1    0        0     1      1      4

###  step 7. 나이브 베이즈 계산을 해봅시다!

#### Q7. 입력문서가 {fast, furious, fun} 을 주요 단어로 가질때, 이 문서는 얼마의 확률로 어떤 문서로 분류가 될까요? ( 계산과정을 채워주세요) 

In [21]:
doc =  np.array(doc, dtype=np.int64)           # dataframe을 np-array로 변환해줍니다.
                                               # return값은 np-matrix가 아니라 np-array입니다.
doc

array([[1, 1, 0, 0, 1, 0, 2, 0, 4],
       [0, 0, 1, 0, 0, 1, 0, 1, 3],
       [1, 1, 1, 1, 2, 0, 0, 0, 5],
       [0, 0, 0, 0, 1, 1, 0, 2, 4],
       [0, 0, 1, 1, 0, 0, 1, 1, 4]])

In [22]:
type(doc)

numpy.ndarray

In [31]:
sum(doc[doc[:,0]==1][:,-1])

9

####  P(Y=comedy), P(Y=action) 계산하기

In [25]:
# P(Y=comedy)
p_comedy = sum(doc[:,0]==1) / len(doc)

# P(Y=action)
p_action = 1-p_comedy

In [26]:
print('p_comedy : ',p_comedy)
print('p_action : ', p_action)

p_comedy :  0.4
p_action :  0.6


#### P(fast=1|comedy=1), P(furious=1|comedy=1), P(fun=1|comedy=1) 계산하기
* 참고 : 문서에 두번 등장한 단어 주의

In [32]:
# P(fast=1|comedy=1)
p_comedy_fast = sum(doc[doc[:,0]==1][:,2]) / sum(doc[doc[:,0]==1][:,-1])

# P(furious=1|comedy=1)
p_comedy_furious = sum(doc[doc[:,0]==1][:,5]) / sum(doc[doc[:,0]==1][:,-1]) 

# P(fun=1|comedy=1)
p_comedy_fun = sum(doc[doc[:,0]==1][:,4]) / sum(doc[doc[:,0]==1][:,-1])

In [33]:
print('p_comedy_fast : ' , p_comedy_fast)
print('p_comedy_furious : ' , p_comedy_furious)
print('p_comedy_fun : ' , p_comedy_fun)   

p_comedy_fast :  0.1111111111111111
p_comedy_furious :  0.0
p_comedy_fun :  0.3333333333333333


#### P(fast=1|action=1), P(furious=1|action=1), P(fun=1|action=1) 계산하기
* 참고 : 문서에 두번 등장한 단어 주의

In [34]:
# P(fast=1|action=1)
p_action_fast = sum(doc[doc[:,0]==0][:,2]) / sum(doc[doc[:,0]==0][:,-1])

# P(furious=1|action=1)
p_action_furious = sum(doc[doc[:,0]==0][:,5]) / sum(doc[doc[:,0]==0][:,-1])

# P(fun=1|action=1)
p_action_fun = sum(doc[doc[:,0]==0][:,4]) / sum(doc[doc[:,0]==0][:,-1])

In [35]:
print('p_action_fast : ' , p_action_fast)
print('p_action_furious : ' , p_action_furious)
print('p_action_fun : ' , p_action_fun)  

p_action_fast :  0.18181818181818182
p_action_furious :  0.18181818181818182
p_action_fun :  0.09090909090909091


#### P(Y = comedy| X = fast, furious, fun) , P(Y = action=1| X = fast, furious, fun) 에서 분자값 계산하기

In [36]:
#P(Y = comedy & X = fast, furious, fun)
proba_comedy = p_comedy_fast * p_comedy_furious * p_comedy_fun * p_comedy 

#P(Y = action=1 & X = fast, furious, fun)
proba_action = p_action_fast * p_action_furious * p_action_fun * p_action 

In [37]:
print('proba_comedy', proba_comedy)
print('proba_action', proba_action)

proba_comedy 0.0
proba_action 0.0018031555221637867


### step 8. 다음 값을 비교해봅시다.
1. P(Y = comedy| X = fast, furious, fun) = 0
2. P(Y = action=1| X = fast, furious, fun) = 0.018


Q8. 어떤 문제점을 발견할수 있나요? 문제점을 해결하기 위한 방법으로는 어떤게 있을까요?
- A. p_comedy_furious가 0이기 때문에 P(Y = comedy| X = fast, furious, fun) 값이 0이 된다.  
확률이 0이면 새로운 데이터가 등장했을 때 예측을 할 수 없으므로 **라플라스 스무딩**을 사용한다.

## + scikit-learn을 이용한 Naive Bayes Classifier 실습

In [38]:
x = np.array(frequency_matrix.drop(['count'], axis=1), dtype=np.int64)
y = np.array(documents['label'],dtype=np.int64)

In [39]:
x

array([[1, 0, 0, 1, 0, 2, 0],
       [0, 1, 0, 0, 1, 0, 1],
       [1, 1, 1, 2, 0, 0, 0],
       [0, 0, 0, 1, 1, 0, 2],
       [0, 1, 1, 0, 0, 1, 1]])

In [40]:
from sklearn.naive_bayes import BernoulliNB
model_bern = BernoulliNB().fit(x,y)

In [41]:
model_bern.classes_ # y클래스의 종류

array([0, 1])

In [42]:
model_bern.class_count_ # y class count

array([3., 2.])

In [43]:
np.exp(model_bern.class_log_prior_) # prior probability

array([0.6, 0.4])

In [44]:
p = model_bern.feature_count_ / np.repeat(model_bern.class_count_[:,np.newaxis], 7, axis=1)
p  # 각 클래스, 특성별로 각각 다른 베르누이 확률변수라고 가정했을 때의 p값

array([[0.        , 0.66666667, 0.33333333, 0.33333333, 0.66666667,
        0.33333333, 1.        ],
       [1.        , 0.5       , 0.5       , 1.        , 0.        ,
        0.5       , 0.        ]])

그러나 위 값은 model 내에서의 값과 다르다. 왜냐면 model에서 구한 값은 smoothing이 된 값이기 때문이다.  
0과 1이 나오는 가상의 데이터를 추가함으로써 추정한 모수의 값이 좀 더 0.5에 가까워지게 만든다.
이 때 사용한 alpha값은 다음과 같다.

In [45]:
model_bern.alpha # alpha = 1 이면 Laplace smoothing (default) / alpha < 1 이면 Lidstone smoothing

1.0

In [46]:
theta = np.exp(model_bern.feature_log_prob_)
theta # smoothing이 적용된 bernoulli 확률분포의 p값 (모수값)

array([[0.2 , 0.6 , 0.4 , 0.4 , 0.6 , 0.4 , 0.8 ],
       [0.75, 0.5 , 0.5 , 0.75, 0.25, 0.5 , 0.25]])

이제 새로운 데이터를 가지고 테스트를 해보려고 한다.

In [47]:
x_new = np.array([0, 1, 0, 1, 1, 0, 0]) # fast, fun, furious가 들어간 text
model_bern.predict_proba([x_new])

array([[0.73898212, 0.26101788]])

action일 확률이 0.74, comedy일 확률이 0.26이다. 즉 action으로 분류될 확률이 0.74이다.